cmd/mistryd/worker_pool.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/skroutz/mistry/pkg/types"
+)
+
+// WorkerPool implements a fixed-size pool of worker goroutines that can be sent
+// build jobs and communicate their result
+type WorkerPool struct {
+	// the fixed amount of goroutines that will be handling running jobs
+	MaxWorkers int
+	// the maximum backlog of pending job requests. if exceeded, sending new work
+	// to the pool will return an error
+	QueueSize int
+	queue     chan workItem
+}
+
+// WorkResult contains the result of a job, either a buildinfo or an error
+type WorkResult struct {
+	BuildInfo *types.BuildInfo
+	Err       error
+}
+
+// FutureWorkResult is a WorkResult that has not yet become available
+type FutureWorkResult struct {
+	resultQueue <-chan WorkResult
+}
+
+// Wait waits for the result to become available and returns it
+func (future FutureWorkResult) Wait() (WorkResult, error) {
+	result, ok := <-future.resultQueue
+	if !ok {
+		return WorkResult{}, fmt.Errorf("Failed to read from result channel")
+	}
+	return result, nil
+}
+
+// workItem contains a job and a channel to place the job result. struct
+// used in the internal work queue
+type workItem struct {
+	job        *Job
+	jobRequest types.JobRequest
+	result     chan<- WorkResult
+}
+
+// NewWorkerPool creates a new worker pool
+func NewWorkerPool(s *Server, maxWorkers, queueSize int, logger *log.Logger) *WorkerPool {
+	p := new(WorkerPool)
+	p.MaxWorkers = maxWorkers
+	p.QueueSize = queueSize
+	p.queue = make(chan workItem, queueSize)
+
+	for i := 0; i < maxWorkers; i++ {
+		go poolWorker(s, i, p.queue)
+	}
+	logger.Printf("Set up %d workers", maxWorkers)
+	return p
+}
+
+// Stop stops the pool workers and closes the queue
+func (p *WorkerPool) Stop() {
+	close(p.queue)
+}
+
+// SendWork sends work to the pool, and receives a work result that can be waited on.
+// An error is returned if the backlog is full and cannot accept any new work items
+func (p *WorkerPool) SendWork(j *Job, jr types.JobRequest) (FutureWorkResult, error) {
```

Comment:
Can we somehow make this agnostic to `Job` and `types.JobRequest`? I think we can. Couldn't this accept a `WorkItem`? This would seem more intuitive and better since we get rid of the tight coupling with the server.